### PR TITLE
Feature/gcp workload monitoring

### PR DESCRIPTION
**Pull request by jcyang43**
# Description

Add option to enable GCP workload monitoring for MaxText workloads. 

- GCP workload monitoring sends performance metrics (heartbeat & training step times) to cloud monarch for monitoring such that if a metric hits its pre-defined threshold, oncalls will be notified to see if any actions are needed. This is ideal for critical workloads sensitive to infrastructure changes. 
- Each metric can be configured to be on or off based on configs. Examples are included in `MaxText/configs/base.yml`
- Documentation can be found at `getting_started/GCP_Workload_Monitoring.md`

# Tests

Tested on trillium TPU and confirmed metrics sent to cloud monarch successfully if configs are enabled. No metrics will be sent to cloud monarch if configs are set to False.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/monitoring/gcp_workload_monitor.py
Comment (gobbleturk) [inline MaxText/monitoring/gcp_workload_monitor.py:1]: This file looks independent of maxtext (other than the maxtext logging plubmed in). Is there any opportunity to move this file / APIs to one of the packages we are also depending on (google-cloud-monitoring, google-api-core
google-api-python-client)

File: MaxText/configs/base.yml [hunk 1]
```diff
@@ -456,6 +456,10 @@ monitor_goodput: True
 goodput_upload_interval_seconds: 60
 enable_pathways_goodput: False
 
+# GCP workload monitoring
+report_heartbeat_metric_for_gcp_monitoring: True
```
Comment (gobbleturk) [inline MaxText/configs/base.yml:460]: Can we default these to false? Most runs of maxtext will not be prod runs that want heartbeating or alerting

File: getting_started/GCP_Workload_Monitoring.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# Enable GCP Workload Monitoring
+This guide provides an overview on how to enable GCP workload monitoring for your MaxText workload.
+
+## Overview
+To address Google Cloud's lack of visibility into user workload performance, users now have the option to enable GCP workload monitoring for all of their MaxText workloads. This workload performance monitoring feature is ideal for critical workloads sensitive to infrastructure changes.
```
Comment (gobbleturk) [inline getting_started/GCP_Workload_Monitoring.md:5]: Can this be worded more positively =)

We offer a monitoring and alerting feature that is well suited for critical workloads sensitive to...

File: requirements.txt [hunk 3]
```diff
@@ -10,6 +10,9 @@ datasets
 gcsfs
 google-cloud-aiplatform==1.61.0
 google-cloud-storage
+google-cloud-monitoring==2.20.0
```
Comment (gobbleturk) [inline requirements.txt:13]: Do these need to be pinned? Can we use latest?

File: MaxText/monitoring/gcp_workload_monitor.py
Comment (jcyang43) [reply MaxText/monitoring/gcp_workload_monitor.py:1]: Hi Matthew, thanks for the suggestion. We're planning to make this improvement in a subsequent version where we'll build an API and python package for workload observability. We'll send a follow-up PR when the API is ready to use

File: getting_started/GCP_Workload_Monitoring.md (hunk 2 above)
Comment (jcyang43) [reply getting_started/GCP_Workload_Monitoring.md]: Done :)

File: requirements.txt (hunk 3 above)
Comment (jcyang43) [reply requirements.txt]: Done

File: MaxText/train.py [hunk 4]
```diff
@@ -852,6 +860,12 @@ def train_loop(config, state=None):
   example_batch = None
   last_step_completion = datetime.datetime.now()
 
+  prof = profiler.Profiler(config)
+  if gcp_workload_monitor and heartbeat_reporting_stop_event:
+    max_logging.log("Starting background thread for reporting heartbeat")
```
Comment (dipannita08) [inline MaxText/train.py:865]: Can we expose an API to start this thread based on configs?

File: MaxText/train.py [hunk 5]
```diff
@@ -868,11 +882,11 @@ def train_loop(config, state=None):
...
 
-    new_time = datetime.datetime.now()
-    record_scalar_metrics(
-        metrics, new_time - last_step_completion, per_device_tflops, learning_rate_schedule(step), per_device_tokens
-    )
-    last_step_completion = new_time
+    step_time_delta = datetime.datetime.now() - last_step_completion
+    record_scalar_metrics(metrics, step_time_delta, per_device_tflops, learning_rate_schedule(step), per_device_tokens)
+    if gcp_workload_monitor and config.report_performance_metric_for_gcp_monitoring:
+      gcp_workload_monitor.report_performance(step_time_delta.total_seconds())
```
Comment (dipannita08) [inline MaxText/train.py:888]: This should ideally be in its own API.

File: MaxText/monitoring/gcp_workload_monitor.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    self.project_id = get_gcp_project_id()
+    self.client = monitoring_v3.MetricServiceClient()
+
+  def report_heartbeat_thread(self, stop_event):
+    """Reports heartbeat metric to GCP every 5 seconds until stop_event is set."""
+    local_rank = os.getenv("LOCAL_RANK", "0")
+    global_rank = jax.process_index()
+    while not stop_event.is_set():
+      self.report_heartbeat(local_rank, str(global_rank))
+      time.sleep(5)
```
Comment (dipannita08) [inline MaxText/monitoring/gcp_workload_monitor.py:35]: Can we make this configurable?

File: MaxText/configs/base.yml (hunk 1 above)
Comment (dipannita08) [reply MaxText/configs/base.yml:463]: Keeping this on by default might lead to errors in internal testing that rely on google3 libraries and some (such as cloud logging and perhaps even monitoring_v3) are not available in google3 yet. MaxText does not support dependency injection in testing so GCP services such as monitoring and logging do not work as intended with tests yet.

I would suggest running some internal tests with the google3 repo as well to validate the above.

**Review by dipannita08 — CHANGES_REQUESTED** [review]
Thank you for this PR!

File: MaxText/train.py (hunk 4 above)
Comment (jcyang43) [reply MaxText/train.py]: done

File: MaxText/train.py (hunk 5 above)
Comment (jcyang43) [reply MaxText/train.py]: done

File: MaxText/monitoring/gcp_workload_monitor.py (hunk 6 above)
Comment (jcyang43) [reply MaxText/monitoring/gcp_workload_monitor.py]: done

**Review by dipannita08 — COMMENTED** [review]
Thanks for the updates! Only recommendation is to turn off monitoring by default until internal testing can be completed.

File: MaxText/configs/base.yml (hunk 1 above)
Comment (jcyang43) [reply MaxText/configs/base.yml]: Defaults updated to false now

**Comment by jcyang43** [comment]
Created new PR without using fork according to submission guidelines: https://github.com/AI-Hypercomputer/maxtext/pull/1167
